### PR TITLE
Refactors paternity calculator flow.

### PR DESCRIPTION
**Pull request by reggieb**
This flow is part of the maternity paternity calculator smart answer.

Some of the refactors were also appied to the adoption calculator flow as the two share a lot of behaviour.

- Replace `save_as_input` with storing response directly on state object
- Replace `calculate` and `precalculate` with assignment within `on_response` blocks
- Remove `precalculate` calls in outcomes and move logic to calculators
- Use `validate` blocks instead of raising `SmartAnswer::InvalidResponse`
- Remove `sprintf` processing for attributes that are processed via `format_money` within erb process.

[Trello card](https://trello.com/c/RpgFrtyo/549-remove-deprecated-flow-methods-from-maternity-paternity-calculator)

### CODE REVIEW

**Review by theseanything — COMMENTED** [review]
Looks good! Could you help me understand where we've move some of the logic for the precalculate methods?

File: lib/smart_answer_flows/maternity-paternity-calculator/adoption_calculator_flow.rb [hunk 1]
```diff
@@ -271,27 +267,7 @@ def define
...
-              end
-              lines.join("\n")
-            end
-          end
-
-          precalculate :total_sap do
-            if above_lower_earning_limit
-              sprintf("%.2f", calculator.total_statutory_pay)
-            end
-          end
```
[the remark this answers is not in the document]
Comment (reggieb) [reply lib/smart_answer_flows/maternity-paternity-calculator/adoption_calculator_flow.rb:288]: The flow uses two calculators: `PaternityPayCalculator` and `AdoptionPayCalculator` and which is assigned to `calculator` depends upon the flow path taken. As both inherit from `MaternityPayCalculator`, the methods are put there to be always available no matter which calculator class is assigned to `calculator`.

File: lib/smart_answer_flows/maternity-paternity-calculator/adoption_calculator_flow.rb [hunk 2]
```diff
@@ -271,27 +267,7 @@ def define
...
-
-          precalculate :average_weekly_earnings do
```
[the remark this answers is not in the document]
Comment (reggieb) [reply lib/smart_answer_flows/maternity-paternity-calculator/adoption_calculator_flow.rb:290]: If you look at where `average_weekly_earnings` is used, it is always passed to `format_money`. So rounding it to two decimal places here is redundant.

File: lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb [hunk 3]
```diff
@@ -463,7 +414,7 @@ def define
         checkbox_question :days_of_the_week_paternity? do
           (0...days_of_the_week.size).each { |i| option i.to_s.to_sym }
 
-          calculate :last_day_in_week_worked do |response|
+          on_response do |response|
```
[the remark this answers is not in the document]
Comment (reggieb) [reply lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb:417]: I can't find anywhere where `last_day_in_week_worked` is used.

File: lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb [hunk 4]
```diff
@@ -520,22 +471,10 @@ def define
...
-
-          precalculate :leave_spp_claim_link do
-            leave_spp_claim_link
-          end
-
-          precalculate :notice_of_leave_deadline do
-            notice_of_leave_deadline
-          end
-
           precalculate :pay_method do
```
[the remark this answers is not in the document]
Comment (reggieb) [reply lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb:474]: We've discussed this before. `pay_method` is defined twice. At [line 300](https://github.com/alphagov/smart-answers/blob/37997f4a93dc440dcb6a39a5ded5a06556547ac0/lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb#L300) it is defined as `calculator.pay_method = calculator.pay_pattern`.

At the moment I have not worked out a way of dealing with this. I've left it as the last thing I fix in the flow - once the rest of the flow has been cleaned up.

To be honest, I was just rather pleased to have sorted out the other `precalculate` within `outcome` calls in this PR. I thought the final "Sort out outcomes" PR was going to be a biggy". Now it's just sorting out this one method.

File: lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb [hunk 5]
```diff
@@ -547,47 +486,8 @@ def define
               end
             )
           end
-
-          precalculate :above_lower_earning_limit do
-            calculator.average_weekly_earnings > calculator.lower_earning_limit
-          end
-
-          precalculate :lower_earning_limit do
-            sprintf("%.2f", calculator.lower_earning_limit)
```
[the remark this answers is not in the document]
Comment (reggieb) [reply lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb:556]: `lower_earning_limit` is passed to template [`_must_earn_over_threshold.erb` where it is passed to `format_money`](https://github.com/alphagov/smart-answers/blob/37997f4a93dc440dcb6a39a5ded5a06556547ac0/lib/smart_answer_flows/maternity-paternity-calculator/outcomes/_must_earn_over_threshold.erb).

File: lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb [hunk 6]
```diff
@@ -547,47 +486,8 @@ def define
...
-          end
-
-          precalculate :entitled_to_pay do
-            above_lower_earning_limit
-          end
```
[the remark this answers is not in the document]
Comment (reggieb) [reply lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb:561]: This is one of the most bizarre bit of code I came across. 

The only place where `entitled_to_pay` is used is on line 564 below where it queried with `above_lower_earning_limit` via an AND operator to determine whether `pay_dates_and_pay` should be constructed from `calculator.paydates_and_pay`.

However, as `entitled_to_pay` is defined as `above_lower_earning_limit` here, the AND operation is redundant.

[the remark this answers is not in the document]
Comment (reggieb) [reply lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb:561]: I expect that in a former iteration, the construction of `entitled_to_pay` was more complicated and therefore the construction made sense then.

File: lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb [hunk 7]
```diff
@@ -547,47 +486,8 @@ def define
...
-          precalculate :total_spp do
-            if above_lower_earning_limit
-              sprintf("%.2f", calculator.total_statutory_pay)
-            end
-          end
-
-          precalculate :average_weekly_earnings do
-            sprintf("%.2f", calculator.average_weekly_earnings)
-          end
-        end
```
[the remark this answers is not in the document]
Comment (reggieb) [reply lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb:581]: Again, these attributes are passed to `format_money` in the erb files.

File: lib/smart_answer_flows/maternity-paternity-calculator/adoption_calculator_flow.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (reggieb) [reply lib/smart_answer_flows/maternity-paternity-calculator/adoption_calculator_flow.rb:288]: To clarify within `smart-answers/lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb` `calculator` is assigned as:

 `self.calculator = Calculators::PaternityPayCalculator.new(due_date)` on [line 26](https://github.com/alphagov/smart-answers/blob/37997f4a93dc440dcb6a39a5ded5a06556547ac0/lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb#L26)

and
`self.calculator = Calculators::PaternityAdoptionPayCalculator.new(matched_date)` on [line 38](https://github.com/alphagov/smart-answers/blob/37997f4a93dc440dcb6a39a5ded5a06556547ac0/lib/smart_answer_flows/maternity-paternity-calculator/paternity_calculator_flow.rb#L38)

**Review by theseanything — APPROVED** [review]

**Pull request merged**